src/Rings/AbelianClosure.jl

```diff
@@ -233,40 +236,51 @@ Hecke.data(a::QQAbFieldElem) = a.data
 #
 ################################################################################
 
-# This function finds a primitive root of unity in our field, note this is
-# not always e^(2*pi*i)/n
+@doc raw"""
+    root_of_unity(K::QQAbField, n::Int)
+
+Return the root of unity $\exp(2\pi i/n)$ as an element of `K`.
+
+# Examples
+```jldocstring
+julia> K, z = abelian_closure(QQ);
 
+julia> root_of_unity(K, 6) == z(6)
+true
+```
+"""
 function root_of_unity(K::QQAbField{AbsSimpleNumField}, n::Int)
+  # Represent the root in the smallest possible cyclotomic field.
   if n % 2 == 0 && n % 4 != 0
     c = div(n, 2)
   else
     c = n
   end
   K, z = cyclotomic_field(K, c)
-  if c == n
-    return QQAbFieldElem{AbsSimpleNumFieldElem}(z, c)
-  else
-    return QQAbFieldElem{AbsSimpleNumFieldElem}(-z, c)
+  if c != n
+    z = -z^div(c+1, 2)
   end
+  return QQAbFieldElem{AbsSimpleNumFieldElem}(z, c)
 end
 
 function root_of_unity(K::QQAbField{AbsNonSimpleNumField}, n::Int)
```

Comment:
Isn't this function identical to the preceding one except for the type parameter? Can't we do this
```suggestion
function root_of_unity(K::QQAbField{T}, n::Int) where T
```
and then just modify the return statement to use `T` to get a single function?